panther_lights/src/controller_node.py

```diff
@@ -174,113 +176,116 @@ def __init__(self, name: str) -> None:
+                animation.front.set_param(req.animation.param)
+                animation.rear.set_param(req.animation.param)
+                animation.reset_time()
+                animation.repeating = req.repeating
+                self._add_animation_to_queue(animation)
+            except ValueError as err:
+                return SetLEDAnimationResponse(False, f'Failed to add animation to queue: {err}')
+
+            return SetLEDAnimationResponse(
+                True, f'Successfully set an animation with id {req.animation.id}'
+            )
 
-    def _set_image_animation_cb(
-        self, req: SetLEDImageAnimationRequest
-    ) -> SetLEDImageAnimationResponse:
-        animation = PantherAnimation()
-        animation.repeating = req.repeating
+    def _set_image_animation_cb(self, req: SetLEDImageAnimationRequest) -> SetLEDImageAnimationResponse:
+        with self._lock:
+            animation = PantherAnimation()
+            animation.repeating = req.repeating
 
-        try:
-            animation_description_front = self._get_image_animation_description(req.front)
-            animation_description_rear = self._get_image_animation_description(req.rear)
+            try:
+                animation_description_front = self._get_image_animation_description(req.front)
+                animation_description_rear = self._get_image_animation_description(req.rear)
 
-            animation.front = BASIC_ANIMATIONS['image_animation'](
-                animation_description_front, self._num_led, self._controller_frequency
-            )
-            animation.rear = BASIC_ANIMATIONS['image_animation'](
-                animation_description_rear, self._num_led, self._controller_frequency
-            )
+                animation.front = BASIC_ANIMATIONS['image_animation'](
+                    animation_description_front, self._num_led, self._controller_frequency
+                )
+                animation.rear = BASIC_ANIMATIONS['image_animation'](
+                    animation_description_rear, self._num_led, self._controller_frequency
+                )
 
-            self._add_animation_to_queue(animation)
-        except Exception as err:
-            return SetLEDImageAnimationResponse(False, f'{err}')
+                self._add_animation_to_queue(animation)
+            except Exception as err:
+                return SetLEDImageAnimationResponse(False, f'{err}')
 
-        return SetLEDImageAnimationResponse(True, f'Successfully set custom animation')
+            return SetLEDImageAnimationResponse(True, f'Successfully set custom animation')
 
     def _update_animations_cb(self, req: TriggerRequest) -> TriggerResponse:
-        try:
-            self._update_user_animations()
-        except Exception as err:
-            return TriggerResponse(False, f'Failed to update animations: {err}')
-        return TriggerResponse(True, 'Animations updated successfully')
+        with self._lock:
```

Comment:
In my opinion, we should only use `_lock` when reading/writing the `BASIC_ANIMATIONS` dictionary. If we use `_lock` only when there is a risk of updating and reading a global variable at the same time, it will be safer and faster. Personally, I don't think it's necessary to protect the entire callback with `_lock`, but ultimately, it's up to you.